File: Assets/Scripts/Attack/Projectile.js
#pragma strict

var cannon : GameObject;

private var damage : float;
private var bonusDamage : float;

var mySpeed : float;
var myRange : float;
var targetPosition : Vector3;
var dx : float;
var dy : float;
var dz : float;

private var myDist : float;

function Start () {
  damage = cannon.GetComponent(Stats).damage;
  var xDis : float = targetPosition.x - transform.position.x;
  var yDis : float = targetPosition.y - transform.position.y;
  var zDis : float = targetPosition.z - transform.position.z;
  var dis : float = Mathf.Sqrt(xDis * xDis + yDis * yDis + zDis * zDis);
  dx = (xDis / dis) * (mySpeed);
  dy = (yDis / dis) * (mySpeed);
  dz = (zDis / dis) * (mySpeed);
}

function Update () {
  transform.position.x += dx * Time.deltaTime;
  transform.position.y += dy * Time.deltaTime;
  transform.position.z += dz * Time.deltaTime;
  myDist += Time.deltaTime * mySpeed;
  if (myDist >= myRange) Destroy(gameObject);
}

function OnCollisionEnter (collision : Collision) {
  if (collision.gameObject.tag == "Enemy") {
    collision.gameObject.GetComponent(Stats).nowHP -= (damage + bonusDamage);
    if (collision.gameObject.GetComponent(Stats).nowHP <= 0) {
      if (Camera.main.GetComponent(MainFunction).learnable && cannon) {
        cannon.GetComponent(Stats).kill++;
        bonusDamage = (cannon.GetComponent(Stats).kill / 10.0) / 100.0 * damage;
      }
    }
    Destroy(gameObject);
  }
}

File: Assets/new/Scripts/Ammunition/Projectile.cs
using UnityEngine;
using System.Collections;

[RequireComponent (typeof(Collider))]
public class Projectile : MonoBehaviour {

  [SerializeField]
  private float shootingTime = 0.2f;
  private float passedTime;
  private bool hasTarget;

  public GameObject explosion;

  private Turret _sourceTurret;
  public Turret SourceTurret {
    get {
      return _sourceTurret;
    }
    set {
      _sourceTurret = value;
    }
  }

  private Vector3 _targetPosition;
  public Vector3 TargetPosition {
    get {
      return _targetPosition;
    }
    set {
      _targetPosition = value;
      hasTarget = true;
    }
  }

  void Update() {
    if (!hasTarget) {
      return;
    }
    transform.position = Vector3.Lerp(transform.position, TargetPosition, passedTime / shootingTime);
    passedTime += Time.deltaTime;
    if (passedTime >= shootingTime) {
      if (explosion != null) {
        GameObject newExplosion = Instantiate(explosion, transform.position, transform.rotation) as GameObject;
        Destroy(newExplosion, 1);
      }
      Destroy(gameObject);
    }
  }

  void OnCollisionEnter(Collision collision) {
    if (collision.gameObject.tag == "Enemy") {
      if (_sourceTurret != null) {
        _sourceTurret.DealDamage(collision.gameObject);
      }
      if (explosion != null) {
        GameObject newExplosion = Instantiate(explosion, transform.position, transform.rotation) as GameObject;
        Destroy(newExplosion, 1);
      }
      Destroy(gameObject);
    }
  }
}
